<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [339]:
import pandas as pd
import psycopg2

In [340]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Данные подключения удалены в целях безопастности


In [341]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [342]:
# текст запроса
query_3_1 = f'''select count(*)
from vacancies
'''

In [343]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1114379196.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [344]:
# текст запроса
query_3_2 = f'''select count(*)
from employers
'''

In [345]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\3062646138.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [346]:
# текст запроса
query_3_3 = f'''select count(*)
from areas
'''

In [347]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\928149064.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [348]:
# текст запроса
query_3_4 = f'''select count(*)
from industries
'''

In [349]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2101691108.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    294

***

In [350]:
# выводы по предварительному анализу данных

Данные распределены по разным таблицам, но их не так много, что-бы были трудности с обработкой и загрузки их всех разом.


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [351]:
# текст запроса
query_4_1 = f'''select a."name"
from areas a
join vacancies v on a.id = v.area_id
group by a.id 
order by count (v.id) desc 
limit 5
'''

In [352]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2457173557.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0           Москва
1  Санкт-Петербург
2            Минск
3      Новосибирск
4           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [353]:
# текст запроса
query_4_2 = f'''select count (*)
from vacancies v 
where salary_from is not null or salary_to is not null
'''

In [354]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\970655580.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [355]:
# текст запроса
query_4_3 = f'''select round(avg(salary_from)) as avg_from, round(avg(salary_to)) as avg_to
from vacancies v 
order by 1
'''

In [356]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2500607024.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [357]:
# текст запроса
query_4_4 = f'''SELECT v.schedule, v.employment, count(v.id)
FROM vacancies v
group by v.schedule , v.employment
order by 3 desc 
offset 1
limit 1
'''


In [358]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2022935265.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [359]:
# текст запроса
query_4_5 = f'''SELECT experience, count (id)
FROM vacancies v
where experience is not null
group by experience
order by count (id)
'''

In [360]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2316260449.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

In [361]:
# визуализация
import plotly
import plotly.express as px
fig = px.pie(df,
            values='count',
            names='experience',
            width=600,
            height=400,
            title='Распределение желаемого опыта работы')
fig.show()

***

In [362]:
# выводы по детальному анализу вакансий

Основная масса вакансий расположена в крупных городах, что и следовало ожидать. Что интересно, работодатели в основном ищут сотрудников с малым опытом работы. Средняя зарплатная вилка - от 70 до 110 тысяч, что является неплохой зарплатой для сотрудников с малым опытом работы.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [363]:
# текст запроса
query_5_1 = f'''SELECT e.name, count (v.id)
FROM vacancies v
join employers e on e.id = v.employer_id 
group by e.name
order by count (v.id) desc
limit 5
'''

In [364]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

In [365]:
#визуализация
fig = px.pie(df,
            values='count',
            names='name',
            width=600,
            height=400,
            title='Распределение работодателей по количеству вакансий')
fig.show()

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [366]:
# текст запроса
query_5_2 = f'''SELECT a.name
--,count(v.id) as vacancy_count
--,count(a.id) as employes_count
from areas a
left join vacancies v on a.id = v.area_id
left join employers e on e.area = a.id
where v.id is null
group by a.id
order by count(v.id),count(a.id) desc
limit 1
'''

In [367]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2250345976.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0  Россия

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [368]:
# текст запроса
query_5_3 = f'''SELECT e.name,
count(distinct v.area_id)
from employers e
left join vacancies v on e.id = v.employer_id 
group by e.id
order by 2 desc
limit 10
'''

In [369]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
5                  Совкомбанк     63
6                         МТС     55
7  ЭФКО, Управляющая компания     49
8                Почта России     48
9                        КРОН     48

In [370]:
#визуализация
fig = px.bar(df,
            x='name',
            y='count',            
            height=400,
            width=700,
            labels={'name':'работодатель','count':'количество регионов'},
            title='Охват регионов работодателями', text_auto=True,
        )
fig.show()

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [371]:
# текст запроса
query_5_4 = f'''select count(*)
from employers e
left join employers_industries ei on e.id = ei.employer_id
where ei.industry_id is null
'''

In [372]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [373]:
# текст запроса
query_5_5 = f'''select e.name
from employers e
left join employers_industries ei on e.id = ei.employer_id
group by e.id
having count(ei.industry_id) = 4
order by 1
limit 1
offset 2
'''

In [374]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [375]:
# текст запроса
query_5_6 = f'''select count(*)
from employers e
left join employers_industries ei on e.id = ei.employer_id
left join industries i on ei.industry_id = i.id
where i.name = 'Разработка программного обеспечения'
'''

In [376]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [377]:
# код для получения списка городов-милионников
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
import requests # Импортируем библиотеку requests

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'

response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')

# найдем таблицу с городами миллионерами
table = soup.find('table', class_ = 'standard sortable')

# извлечем названия городов из таблицы
cities = []
for row in table.find_all('tr')[1:]:
    cells = row.find_all('td')
    city = cells[1].text.strip()
    city = city.replace("[a]","") # убираем сноску из Воронежа
    cities.append(city)

cities=tuple(cities)
print(cities)

# не уверен, что  надо было сделать именно так. По моему - это сложно для задачи, по которой мы прошлись очень вскользь, и практически не практиковались.

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


In [378]:
# текст запроса
query_5_7 = f'''select a.name, count(a.id)
from employers e
join vacancies v on v.employer_id = e.id
join areas a on v.area_id = a.id
where e.name = 'Яндекс' and a.name in {cities}
group by a.id
union all
select 'Total', count(a.id)
from employers e
join vacancies v on v.employer_id = e.id
join areas a on v.area_id = a.id
where e.name = 'Яндекс' and a.name in {cities}
order by 2
'''

In [379]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7            Самара     26
8               Уфа     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

In [380]:
# выводы по анализу работодателей

Крупнейший работодатель на этом рынке - компания Яндекс. Она - абсолютный лидер, как по количеству вакансий, так и по широте географического охвата.
Огромное количество компаний не указывают свою сферу деятельности. Потенциально, это может отпугнуть возможных соискателей.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [381]:
# текст запроса
query_6_1 = f'''select count(*)
from vacancies v
where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [382]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [383]:
# текст запроса
query_6_2 = f'''select count(*)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
and
(lower(v.name) like '%junior%' or
v.experience = 'Нет опыта' or
v.employment = 'Стажировка')
'''

In [384]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
jun_count=df['count'][0]
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\691433469.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0     51

In [385]:
#визуализация
query_6_3_1 = f'''select 
count(*)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
'''
df = pd.read_sql_query(query_6_3_1, connection)
all_count=df['count'][0]

fig = px.pie(pd.DataFrame([jun_count,all_count- jun_count], index=['sql','other'], columns=['num']),
            values='num',            
            names=['Требуется Junior','Все остальные'],
            width=500,
            height=300,
            title='Доля поиска Junior-специалиста среди вакансий')
fig.show()

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2034595520.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [386]:
# текст запроса
query_6_3 = f'''select 
count(*)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
and
(
lower(v.key_skills) like '%sql%' or
lower(v.key_skills) like '%postgres%'
)
'''

In [387]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
sql_count=df['count'][0]
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\652152469.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    201

In [388]:
#визуализация

fig = px.pie(pd.DataFrame([sql_count,all_count- sql_count], index=['sql','other'], columns=['num']),
            values='num',            
            names=['Требуется SQL/Postgres','Не указано'],
            width=500,
            height=300,
            title='Доля требования SQL/Postgres в вакансиях')
fig.show()

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [389]:
# текст запроса
query_6_4 = f'''select 
count(*)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
and
(
lower(v.key_skills) like '%python%'
)
'''

In [390]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
py_count=df['count'][0]
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1493346929.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    351

In [391]:
#визуализация
fig = px.pie(pd.DataFrame([py_count,all_count - py_count], index=['py','other'], columns=['num']),
            values='num',            
            names=['Требуется Python','Не указано'],
            width=500,
            height=300,
            title='Доля требования Python в вакансиях')
fig.show()

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [392]:
# текст запроса
query_6_5 = f'''select 
round(avg(LENGTH(key_skills) - LENGTH(replace (key_skills,CHR(9) ,''))+1),2)
--LENGTH(key_skills) - LENGTH(replace (key_skills,CHR(9) ,''))+1
--,key_skills
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
and
v.key_skills is not null

'''

In [393]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [394]:
# текст запроса
query_6_6 = '''select 
experience,
round(avg((coalesce(salary_from,salary_to) + coalesce(salary_to,salary_from)) / 2)) as salary_avg
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
and
(
v.salary_from is not null or v.salary_to is not null
)
group by v.experience
order by 2
'''

In [395]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  salary_avg
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

In [396]:
#визуализация
fig = px.bar(df,
            x='experience',
            y='salary_avg',
            height=300,
            width=600,
            labels={'experience':'Опыт работы','salary_avg':'Средняя зарплата'},
            title='Средний уровень зарплаты в зависимости от опыта'
        )
fig.show()

***

In [397]:
# выводы по предметному анализу

DS - выглядит весьма переспективным направлением, в плане зарплаты и роста. За довольно короткий промежуток времени зарплата DS-специалиста вырастает более, чем в 2 раза.<br>
Тем не менее, требования к DS специалисту довольно высоки. Подавляющее большинство вакансий (больше 3/4) требует знания Python, и больше половины вакансий указывают в требованиях SQL/Postgres.<br>
10% вакансий составляет поиск молодого специалиста, Junior. Сложно сказать, является-ли это большой, или наоборот малой величиной. Для этого потребуется сравнить с другими специализациями.<br>
В базе очень мало вакансий, где требовалось опыт работы больше 6 лет. Что неудивительно - профессия очень молодая, и людей с большим опытом работы просто не успело появиться.


# Общий вывод по проекту

In [398]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [399]:
query = '''select 
v.experience , count(v.id)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
group by experience
order by 2 desc
'''
df = pd.read_sql_query(query, connection)
fig1 = px.pie(df,
            values='count',
            names='experience',
            width=630,
            height=300,
            title='Распределение желаемого опыта работы среди DS-специалистов')
fig1.show()
query = '''select 
v.experience , count(v.id)
from vacancies v
group by experience
order by 2 desc
'''
df = pd.read_sql_query(query, connection)
fig2 = px.pie(df,
            values='count',
            names='experience',
            width=630,
            height=300,
            title='Распределение желаемого опыта работы среди IT-специалистов')

fig2.show()

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\322528417.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\322528417.py:30: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Как мы видим, DS-специалистов предпочитают брать хотя-бы с минимальным опытом работы, хотя-бы в 1год, по сравнению с IT-специалистами.<br>
Это может свидетельстовать о том, что новичку может быть сложнее найти работу в специальности DS, чем в других IT специальностях.<br>
Конечно, это зависит еще и от предложения на рынке труда, но у нас пока нет этих данных, что-бы сделать такой анализ

In [400]:
query = '''select 
replace (name, 'C','С') as nam
, count (v.id)
from vacancies v 
group by nam
order by 2 desc
limit 10
'''
df = pd.read_sql_query(query, connection)
fig = px.pie(df,
            values='count',
            names='nam',
            width=600,
            height=400,
            title='Распределение IT вакансий на рынке')
fig.show()

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\2445128942.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



На рынке подавляющее большинство вакансий занимает 'Системный администратор' и 'Программист 1С'.
Хотя, возможно, это связано с тем, что специальности, давно существующие на рынке имеют свои, устоявшися названия. В то время как, например, DS-специалиста могут называть по разному.

In [401]:
query = '''select 
lower(name) as name, count (v.id)
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
group by lower(v.name)
order by 2 desc
limit 10
'''
df = pd.read_sql_query(query, connection)
print (df)

query = '''select 
--lower(name), count (v.id)
--count(*)
count (distinct lower(v."name"))
from vacancies v
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)'''
df = pd.read_sql_query(query, connection)
print ('Всего вариантов специальности',df)

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1437439399.py:17: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



                                                name  count
0                                     data scientist     63
1                              senior data scientist     23
2                              c++ разработчик (qml)     10
3                  ml engineer/ data scientist (waf)      9
4  junior data scientist (валидация скоринговых м...      9
5                              middle data scientist      8
6                                        ml engineer      6
7              архитектор решений с опытом в tiny ml      6
8                          machine learning engineer      6
9          python-разработчик в команду data science      5


C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1437439399.py:34: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Всего вариантов специальности    count
0    277


Как мы видим, чаще всего в вакансии указывают просто - Data scientist. 
Однако всего, различных вариаций написания специальносей в этом направлении - 277. Это довольно много, и может серьезно осложнить поиск соискателем нужной вакансии.

In [403]:
query = '''select 
a.name,
count (v.id)
from vacancies v
join areas a on v.area_id = a.id
join employers e on v.employer_id = e.id 
--where a."name" not in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
group by a.id
order by 2 desc 
limit 10
'''
df = pd.read_sql_query(query, connection)
print (df)

              name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892
5     Екатеринбург   1698
6  Нижний Новгород   1670
7           Казань   1415
8        Краснодар   1301
9           Самара   1144


C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1397291460.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Распределение вакансий по городам - довольно ожидаемое. Чем крупнее город - тем больше вакансий. Больше всего вакансий, разумеется, в городах-миллионниках.

In [404]:
query = '''select 
a.name,
count (v.id)
from vacancies v
join areas a on v.area_id = a.id
join employers e on v.employer_id = e.id
where
(
lower(v.name) like '%data scientist%' or
lower(v.name) like '%data science%' or
lower(v.name) like '%исследователь данных%' or
lower(v.name) like '%machine learning%' or
lower(v.name) like '%машинн%обучен%' or
(v.name like '%ML%' and v.name not like '%HTML%')
)
group by a.id
order by 2 desc 
limit 10
'''
df = pd.read_sql_query(query, connection)
print (df)

C:\Users\trigubov\AppData\Local\Temp\ipykernel_11948\1113248042.py:20: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



              name  count
0           Москва    217
1  Санкт-Петербург     64
2      Новосибирск     23
3  Нижний Новгород     20
4           Алматы     16
5           Казань     16
6            Минск      9
7            Томск      8
8     Екатеринбург      7
9          Армения      6


А вот с распределением вакансий DS-специалистов уже иная картина. Подавляющее большинство вакансий сосредоточено в Москве. Скорее всего это связано с тем, что Москва - более высокотехнологичный центр, чем другие регионы.

В целом, базе данных вакансий HH явно не хватает какой-то структиризации профессий. Одна и та-же профессия может называться десятками разных вариантов. Даже одно и тоже написание может быть разным. Например "1С программист" и "Программист 1С", более того, буква 'C' в базе встречается как в английской транскрипции, так и в русской.

Я-бы попробовал рассмотреть внедрение "облака тегов" для вакансий. Возможно, это позволит лучше структирировать информацию, и серьезно облегчить поиск нужной вакансии.